File: leetcode/leetcode_17.py
class Solution:
    def letterCombinations(self, digits):
        digitList = list(map(int, list(digits)))
        graph = [[], []]
        answer = []
        cnt = 0
        
        for i in range(1, 9):
            plus = 4 if i == 6 or i == 8 else 3
            graph.append(list(chr(s) for s in range(97 + cnt, 97 + cnt + plus)))
            cnt += plus
        
        def combi(index, str = ''):
            if not digitList:
                return
            if len(str) == len(digitList):
                answer.append(str)
                return
            for char in graph[digitList[index]]:
                combi(index + 1, str + char)

        combi(0)
        return answer;

File: leetcode/leetcode_17.js
const letterCombinations = (digits) => {
  const digitList = digits.split('').map(Number);
  const graph = [[], []];
  const answer = [];
  let cnt = 0;

  for (let i = 2; i < 10; i++) {
    let plus = i === 7 || i === 9 ? 4 : 3;
    let alpha = '';
    for (let j = 97 + cnt; j < 97 + cnt + plus; j++) {
      alpha += String.fromCharCode(j);
    }
    graph.push(alpha.split(''));
    cnt += plus;
  }

  const combi = (index, str = '') => {
    if (!digits) {
      return;
    }
    if (str.length == digitList.length) {
      answer.push(str);
      return;
    }
    graph[digitList[index]].forEach((char) => {
      combi(index + 1, str + char);
    });
  };

  combi(0);
  return answer;
};
